dynamic/discovery/discovery.go

```diff
@@ -139,34 +135,24 @@ func (rm *ResourceMap) refresh() {
 			// Remember how to map back from Kind/subresource to resource.
 			// This is different from what RESTMapper provides because we already know
 			// the full GroupVersionKind and just need the resource name.
+			// Remember which resources are subresources, and map the kind to the main resource.
 			if strings.ContainsRune(apiResource.Name, '/') {
-				// Remember which resources are subresources, and map the kind to the main resource.
 				gve.subresources[apiResource.Name] = apiResource
 			} else {
-				// Make sure we don't choose a subresource like "pods/status".
 				gve.kinds[apiResource.Kind] = apiResource
 			}
 		}
 
 		// Group all subresources for a resource.
-		for i := range group.APIResources {
-			apiResource := &APIResource{
-				APIResource: group.APIResources[i],
-				APIVersion:  group.GroupVersion,
-			}
-			if !strings.ContainsRune(apiResource.Name, '/') {
-				apiResource.SubresourceMap = make(map[string]bool)
-				for _, apiSubresource := range gve.subresources {
-					split := strings.Split(apiSubresource.Name, "/")
-					resourceName := split[0]
-					subresourceKey := split[1]
-					if apiResource.Name == resourceName {
-						apiResource.SubresourceMap[subresourceKey] = true
-					}
-				}
-				gve.resources[apiResource.Name] = apiResource
-				gve.kinds[apiResource.Kind] = apiResource
+		for apiSubresourceName, _ := range gve.subresources {
+			arr := strings.Split(apiSubresourceName, "/")
+			apiResourceName := arr[0]
+			subresourceKey := arr[1]
+			apiResource := gve.resources[apiResourceName]
+			if len(apiResource.subresourceMap) == 0 {
```

Comment:
This looks awkward, but it saves unnecessary initialization of `.subresourceMap` for subresources by not initializing the field for all resources upfront in previous loop.

Reply:
I think the intent is more clear if we compare with `nil`. We don't really care whether there are any items. We care if the map has been allocated or not.

```suggestion
			if apiResource.subresourceMap == nil {
```